tmva/tmva/inc/TMVA/DNN/Minimizers.h

```diff
@@ -323,43 +295,33 @@ auto TGradientDescent<Architecture_t>::TrainMomentum(const Data_t & trainingData
        }
    }
 
-   std::chrono::time_point<std::chrono::system_clock> start, end;
-   start = std::chrono::system_clock::now();
-
-   while (!converged)
-   {
-      fStepCount++;
-
-      trainLoader.Shuffle();
-      // Iterate over epoch.
-      std::vector<TBatch<Architecture_t>> batches{};
-      for (size_t i = 0; i < nTrainingSamples / net.GetBatchSize(); i += nThreads) {
-         batches.clear();
-         batches.reserve(nThreads);
-         for (size_t j = 0; j < nThreads; j++) {
-            batches.push_back(trainLoader.GetBatch());
-         }
-         if (momentum != 0.0) {
-            StepMomentum(net, nets, batches, momentum);
-         } else {
-            Step(net, nets, batches);
+   size_t batchesInEpoch = nTrainingSamples / net.GetBatchSize();
+   std::vector<TBatch<Architecture_t>> batches{};
+   batches.reserve(nThreads);
+
+   do {
+      for (fStepCount = 0; fStepCount < fTestInterval; fStepCount++) {
+         trainLoader.Shuffle();
+         for (size_t i = 0; i < batchesInEpoch; i += nThreads) {
```

Comment:
This is not introduced by you but: what happens for `nThreads == 4`, `nTrainingSamples == 7`? What does `batches.push_back()` do below, for `i == 1` and ` j == 3`?